src/charm.py

```diff
@@ -109,13 +121,41 @@ def __init__(self, *args):
             self._certificates.on.certificate_expiring, self._on_certificate_expiring
         )
 
+    def _configure_nrf(self, _: EventBase) -> None:
+        """Handle Juju events.
+
+        This event handler is called for every event that affects the charm state
+        (ex. configuration files, relation data). This method performs a couple of checks
+        to make sure that the workload is ready to be started. Then, it generates a configuration
+        for the NRF workload, runs the Pebble services and exposes service information
+        the requirers.
+        """
+        if not self.ready_to_configure():
+            logger.info("The preconditions for the configuration are not met yet.")
+            return
+        if not self._private_key_is_stored():
+            self._generate_private_key()
+        if not self._csr_is_stored():
+            self._request_new_certificate()
+        provider_certificate = self._get_current_provider_certificate()
+        if not provider_certificate:
+            return
+        if certificate_update_required := self._is_certificate_update_required(
+                provider_certificate
+        ):
+            self._store_certificate(certificate=provider_certificate)
+        desired_config_file = self._generate_nrf_config_file()
+        if config_update_required := self._is_config_update_required(desired_config_file):
+            self._push_config_file(content=desired_config_file)
+        self._configure_workload(restart=(config_update_required or certificate_update_required))
+        self._publish_nrf_info_for_all_requirers()
+
     def ready_to_configure(self) -> bool:
         """Return whether all preconditions are met to proceed with configuration."""
         if not self._container.can_connect():
             return False
-        for relation in [DATABASE_RELATION_NAME, "certificates"]:
-            if not self._relation_created(relation):
-                return False
+        if self._missing_relations():
+            return False
         if not self._database_is_available():
```

Comment:
We can also check here whether Webui data is available or not.

Reply:
That's right! I had a feeling that something was missing :)
Added.